src/number_theory/cyclotomic/basic.lean

```diff
@@ -0,0 +1,225 @@
+  refine ⟨λ h, ⟨((iff _ _ _).1 h).1 n (set.mem_singleton _), by simpa using ((iff _ _ _).1 h).2⟩,
+  λ h, ⟨λ a ha, _, by simpa using h.2⟩⟩,
+  obtain ⟨⟨b, hb⟩, H⟩ := h,
+  exact ⟨b, (set.mem_singleton_iff.1 ha).symm ▸ hb⟩
+end
+
+/-- If `is_cyclotomic_extension ∅ A B`, then `A = B`. -/
+lemma empty [h : is_cyclotomic_extension ∅ A B] : (⊤ : subalgebra A B) = ⊥ :=
+begin
+  replace h := h.adjoint_roots,
+  simp only [set.mem_empty_eq, set.set_of_false, adjoin_empty, exists_false, false_and] at h,
+  exact (algebra.eq_top_iff.2 h).symm,
+end
+
+/-- Transitivity of cyclotomic extensions. -/
+lemma trans (C : Type w) [comm_ring C] [algebra A C] [algebra B C]
+  [is_scalar_tower A B C] [hS : is_cyclotomic_extension S A B]
+  [hT : is_cyclotomic_extension T B C] : is_cyclotomic_extension (S ∪ T) A C :=
+begin
+  refine ⟨λ n hn, _, λ x, _⟩,
+  { cases hn,
+    { obtain ⟨b, hb⟩ := ((iff _ _ _).1 hS).1 n hn,
+      refine ⟨algebra_map B C b, _⟩,
+      replace hb := congr_arg (algebra_map B C) hb,
+      rw [aeval_def, eval₂_eq_eval_map, map_cyclotomic, ring_hom.map_zero, ← eval₂_at_apply,
+        eval₂_eq_eval_map, map_cyclotomic] at hb,
+      rwa [aeval_def, eval₂_eq_eval_map, map_cyclotomic] },
+    { obtain ⟨c, hc⟩ := ((iff _ _ _).1 hT).1 n hn,
+      refine ⟨c, _⟩,
+      rw [aeval_def, eval₂_eq_eval_map, map_cyclotomic] at hc,
+      rwa [aeval_def, eval₂_eq_eval_map, map_cyclotomic] } },
+  { refine adjoin_induction (((iff _ _ _).1 hT).2 x) (λ c ⟨n, hn⟩, subset_adjoin
+      ⟨n, or.inr hn.1, hn.2⟩) (λ b, _) (λ x y hx hy, subalgebra.add_mem _ hx hy)
+      (λ x y hx hy, subalgebra.mul_mem _ hx hy),
+    { let f := is_scalar_tower.to_alg_hom A B C,
+      have hb : f b ∈ (adjoin A { b : B | ∃ (a : ℕ+), a ∈ S ∧ b ^ (a : ℕ) = 1 }).map f :=
+        ⟨b, ((iff _ _ _).1 hS).2 b, rfl⟩,
+      rw [is_scalar_tower.to_alg_hom_apply, ← adjoin_image] at hb,
+      refine adjoin_mono (λ y hy, _) hb,
+      obtain ⟨b₁, ⟨⟨n, hn⟩, h₁⟩⟩ := hy,
+      exact ⟨n, ⟨mem_union_left T hn.1, by rw [← h₁, ← alg_hom.map_pow, hn.2, alg_hom.map_one]⟩⟩ } }
+end
+
+/-- If `B` is a cyclotomic extension of `A` given by roots of unity of order in `S ∪ T`, then `B`
+is a cyclotomic extension of `adjoin A { b : B | ∃ a : ℕ+, a ∈ S ∧ b ^ (a : ℕ) = 1 } ` given by
+roots of unity of order in `T`. -/
+lemma union_right [h : is_cyclotomic_extension (S ∪ T) A B]
+  : is_cyclotomic_extension T (adjoin A { b : B | ∃ a : ℕ+, a ∈ S ∧ b ^ (a : ℕ) = 1 }) B :=
+begin
+  have : { b : B | ∃ (n : ℕ+), n ∈ S ∪ T ∧ b ^ (n : ℕ) = 1 } =
+    { b : B | ∃ (n : ℕ+), n ∈ S ∧ b ^ (n : ℕ) = 1 } ∪
+    { b : B | ∃ (n : ℕ+), n ∈ T ∧ b ^ (n : ℕ ) = 1 },
+  { refine le_antisymm (λ x hx, _) (λ x hx, _),
+    { obtain ⟨n, hn⟩ := hx,
+      cases hn.1 with h₁ h₂,
+      { left, exact ⟨n, ⟨h₁, hn.2⟩⟩ },
+      { right, exact ⟨n, ⟨h₂, hn.2⟩⟩ } },
+    { cases hx,
+      { obtain ⟨n, hn⟩ := hx,
+        exact ⟨n, ⟨or.inl hn.1, hn.2⟩⟩ },
+      { obtain ⟨n, hn⟩ := hx,
+        exact ⟨n, ⟨or.inr hn.1, hn.2⟩⟩ } } },
```

Comment:
I guess you could golf this even more using `rcases hx with ⟨n, hn⟩ | ⟨n, hn⟩,` if you want